File: A121_Range_Sensor_Demo/Arduino/A121_Presence_Detector/A121_Presence_Detector.cpp
#include "A121_Presence_Detector.h"

void A121_I2C_Write_Byte(uint16_t reg,uint32_t data)
{
  Wire.beginTransmission(A121_DEV_ADDR);
  Wire.write((reg >> 8) & 0xff);
  Wire.write(reg & 0xff);
  Wire.write((data >> 24) & 0xff);
  Wire.write((data >> 16) & 0xff);
  Wire.write((data >> 8) & 0xff);
  Wire.write(data  & 0xff);
  Wire.endTransmission();
}

void A121_I2C_Read_Byte(uint16_t reg,uint32_t *data)
{
  Wire.beginTransmission(A121_DEV_ADDR);
  Wire.write((reg >> 8) & 0xff);
  Wire.write(reg & 0xff);
  uint8_t error = Wire.endTransmission(true);
  // Serial.printf("endTransmission: %u\n", error);

  //Read 16 bytes from the slave
  uint8_t bytesReceived = Wire.requestFrom(A121_DEV_ADDR, 4);
  // Serial.printf("requestFrom: %u\n", bytesReceived);
  if ((bool)bytesReceived) {  //If received more than zero bytes
    uint8_t temp[bytesReceived];
    Wire.readBytes(temp, bytesReceived);
    *data = ((uint32_t)temp[0] << 24) | ((uint32_t)temp[1] << 16)  | ((uint32_t)temp[2] << 8)  | (uint32_t)temp[3];
  }
}

void A121_Set_Start_And_End_Range(uint32_t start,uint32_t end)
{
  A121_I2C_Write_Byte(REG_START, start);
  A121_I2C_Write_Byte(REG_END, end);
}

void A121_Set_Sweeps_Per_Frame(uint32_t value)
{
  A121_I2C_Write_Byte(REG_SWEEPS_PER_FRAME, value);
}

void A121_Set_Inter_Frame_Presence_Timeout(uint32_t value)
{
  A121_I2C_Write_Byte(REG_INTER_FRAME_PRESENCE_TIMEOUT, value);
}

void A121_Set_Intra_Detection_Enabled(uint32_t value)
{
  A121_I2C_Write_Byte(REG_INTRA_DETECTION_ENABLED, value);
}

void A121_Set_Inter_Detection_Enabled(uint32_t value)
{
  A121_I2C_Write_Byte(REG_INTER_DETECTION_ENABLED, value);
}

void A121_Set_Frame_Rate(uint32_t value)
{
  A121_I2C_Write_Byte(REG_FRAME_RATE, value);
}

void A121_Set_Intra_Detection_Threshold(uint32_t value)
{
  A121_I2C_Write_Byte(REG_INTRA_DETECTION_THRESHOLD, value);
}

void A121_Set_Inter_Detection_Threshold(uint32_t value)
{
  A121_I2C_Write_Byte(REG_INTER_DETECTION_THRESHOLD, value);
}

void A121_Set_Inter_Frame_Deviation_Time_Const(uint32_t value)
{
  A121_I2C_Write_Byte(REG_INTER_FRAME_DEVIATION_TIME_CONST, value);
}

void A121_Set_Inter_Frame_Fast_Cutoff(uint32_t value)
{
  A121_I2C_Write_Byte(REG_INTER_FRAME_FAST_CUTOFF, value);
}

void A121_Set_Inter_Frame_Slow_Cutoff(uint32_t value)
{
  A121_I2C_Write_Byte(REG_INTER_FRAME_SLOW_CUTOFF, value);
}

void A121_Set_Intra_Frame_Time_Const(uint32_t value)
{
  A121_I2C_Write_Byte(REG_INTRA_FRAME_TIME_CONST, value);
}

void A121_Set_Intra_Output_Time_Const(uint32_t value)
{
  A121_I2C_Write_Byte(REG_INTRA_OUTPUT_TIME_CONST, value);
}

void A121_Set_Inter_Output_Time_Const(uint32_t value)
{
  A121_I2C_Write_Byte(REG_INTER_OUTPUT_TIME_CONST, value);
}

void A121_Set_Auot_Profile_Enabled(uint32_t value)
{
  A121_I2C_Write_Byte(REG_AUTO_PROFILE_ENABLED, value);
}

void A121_Set_Auot_Step_Length_Enabled(uint32_t value)
{
  A121_I2C_Write_Byte(REG_AUTO_STEP_LENGTH_ENABLED, value);
}

void A121_Set_Manual_Profile(uint32_t value)
{
  A121_I2C_Write_Byte(REG_MANUAL_PROFILE, value);
}

void A121_Set_Manual_Step_Length(uint32_t value)
{
  A121_I2C_Write_Byte(REG_MANUAL_STEP_LENGTH, value);
}

void A121_Set_Reset_Filters_On_Prepare(uint32_t value)
{
  A121_I2C_Write_Byte(REG_RESET_FILTERS_ON_PREPARE, value);
}

void A121_Set_Hwaas(uint32_t value)
{
  A121_I2C_Write_Byte(REG_HWAAS, value);
}

void A121_Set_Auotmatic_Subsweeps(uint32_t value)
{
  A121_I2C_Write_Byte(REG_AUTOMATIC_SUBSWEEPS, value);
}

void A121_Set_Signal_Quality(uint32_t value)
{
  A121_I2C_Write_Byte(REG_SIGNAL_QUALITY, value);
}

void A121_Set_Command(uint32_t value)
{
  A121_I2C_Write_Byte(REG_COMMAND, value);
}

void A121_Get_Version(uint32_t *value)
{
  A121_I2C_Read_Byte(REG_VERSION, value);
  Serial.print("Major: 0x");
  Serial.print((*value >> 16) & 0xFF, HEX);
  Serial.print(" Minor: 0x");
  Serial.print((*value >> 8) & 0xFF, HEX);
  Serial.print(" Patch: 0x");
  Serial.println(*value & 0xFF, HEX);
}

void A121_Get_Protocol_Status(uint32_t *value)
{
  A121_I2C_Read_Byte(REG_PROTOCOL_STATUS, value);
}

void A121_Get_Measure_Counter(uint32_t *value)
{
  A121_I2C_Read_Byte(REG_MEASURE_COUNTER, value);
}

void A121_Get_Detector_Status(uint32_t *value)
{
  A121_I2C_Read_Byte(REG_DETECTOR_STATUS, value);
}

void A121_Get_Presence_Result(uint32_t *value)
{
  A121_I2C_Read_Byte(REG_PRESENCE_RESULT, value);
}

void A121_Get_Presence_Distance(uint32_t *value) //Monitoring the distance at which an object exists
{
  A121_I2C_Read_Byte(REG_PRESENCE_DISTANCE, value);
}

void A121_Get_Intra_Presence_Score(uint32_t *value)//Fast movement range
{
  A121_I2C_Read_Byte(REG_INTRA_PRESENCE_SCORE, value);
}

void A121_Get_Inter_Presence_Score(uint32_t *value)//Slow-motion amplitude
{
  A121_I2C_Read_Byte(REG_INTER_PRESENCE_SCORE, value);
}

void A121_Get_Presence_Actual_Frame_Rate(uint32_t *value) //Actual frame rate
{
  A121_I2C_Read_Byte(REG_PRESENCE_ACTUAL_FRAME_RATE, value);
}

void A121_Get_Application_Id(uint32_t *value)
{
  A121_I2C_Read_Byte(REG_APPLICATION_ID, value);
}

void A121_init()
{
  Serial.begin(115200);

  pinMode(A121_WAKEUP_PIN, OUTPUT);
  pinMode(A121_BUSY_PIN, INPUT);

  digitalWrite(A121_WAKEUP_PIN, 1);

  Wire.setClock(400000);
  Wire.begin();

  uint32_t value;
  while (digitalRead(A121_BUSY_PIN) == 0);

  A121_Get_Application_Id(&value);
  Serial.print("A121_Get_Application_Id: 0x");
  Serial.println(value, HEX);

  A121_Set_Command(RESET_MODULE);
  delay(100);

  A121_Set_Command(ENABLE_UART_LOGS);

  while (digitalRead(A121_BUSY_PIN) == 0);

  A121_Set_Start_And_End_Range(300, 2500);
  A121_Set_Frame_Rate(12 * 1000);
  A121_Set_Sweeps_Per_Frame(16);

  if (AUTOMATIC_SUBSWEEPS)
  {
    A121_Set_Auotmatic_Subsweeps(true);
    A121_Set_Signal_Quality(20 * 1000);
  }
  else
  {
    A121_Set_Auotmatic_Subsweeps(true);
    A121_Set_Hwaas(32);
    if (OVERRIDE)
    {
      A121_Set_Manual_Profile(PROFILE1);
      A121_Set_Manual_Step_Length(24);
    }
  }

  if (INTRA_MOTION_DETECTION)
  {
    A121_Set_Intra_Detection_Threshold(1300);
    A121_Set_Intra_Frame_Time_Const(150);
    A121_Set_Intra_Output_Time_Const(300);
  }

  if (INTER_MOTION_DETECTION)
  {
    A121_Set_Inter_Detection_Threshold(1000);
    A121_Set_Inter_Frame_Fast_Cutoff(6000);
    A121_Set_Inter_Frame_Slow_Cutoff(200);
    A121_Set_Inter_Frame_Deviation_Time_Const(500);
    A121_Set_Inter_Output_Time_Const(2000);
    A121_Set_Inter_Frame_Presence_Timeout(3);
  }

  A121_Set_Command(APPLY_CONFIGURATION);

  while (1)
  {
    A121_Get_Detector_Status(&value);
    Serial.print("A121_Get_Detector_Status: 0x");
    Serial.println(value, HEX);

    if ((value & ALL_ERROR) == 0)
      break;
    delay(100);
  }

  Serial.println("A121 init OK");
}


void A121_Get_Presence_Detector()
{
  uint32_t value;
  A121_Set_Command(START_DETECTOR);

  while (1)
  {
    A121_Get_Detector_Status(&value);
    // Serial.print("A121_Get_Detector_Status: 0x");
    // Serial.println(value, HEX);
    if ((value & ALL_ERROR) == 0)
      break;
    delay(10);
  }

  A121_Get_Presence_Result(&value);

  if ((value & RESULT_DETECTOR_ERROR) == 0)
  {
    if (value & PRESENCE_DETECTED)
    {
      A121_Get_Presence_Distance(&value);
      Serial.print("Presence_Distance: ");
      Serial.print(value);
      Serial.println(" mm");

      A121_Get_Intra_Presence_Score(&value); // 快速动作幅度
      Serial.print("Intra_Presence_Score: ");
      Serial.println(value);

      A121_Get_Inter_Presence_Score(&value); // 慢动作幅度
      Serial.print("Inter_Presence_Score: ");
      Serial.println(value);

      A121_Get_Presence_Actual_Frame_Rate(&value); // 实际帧率
      Serial.print("Actual_Frame_Rate: ");
      Serial.println(value);
      Serial.println();
    }
    else
    {
      if (value & PRESENCE_DETECTED_STICKY)
      {
        Serial.println("PRESENCE_DETECTED_STICKY");
      }
    }
  }
  else
  {
    Serial.println("A121 DETECTOR ERROR"); // 需要 RESET_MODULE
  }
}

File: A121_Range_Sensor_Demo/ESP32S3/micropython/A121_Presence_Detector.py
import time
from machine import I2C, Pin

# ===============================
# Feature switches
# ===============================
AUTOMATIC_SUBSWEEPS     = True
OVERRIDE                = True
INTRA_MOTION_DETECTION  = True
INTER_MOTION_DETECTION  = True

# ===============================
# Register map
# ===============================
REG_VERSION                          = 0x0000
REG_PROTOCOL_STATUS                  = 0x0001
REG_MEASURE_COUNTER                  = 0x0002
REG_DETECTOR_STATUS                  = 0x0003

REG_PRESENCE_RESULT                  = 0x0010
REG_PRESENCE_DISTANCE                = 0x0011
REG_INTRA_PRESENCE_SCORE             = 0x0012
REG_INTER_PRESENCE_SCORE             = 0x0013
REG_PRESENCE_ACTUAL_FRAME_RATE       = 0x0020

REG_SWEEPS_PER_FRAME                 = 0x0040
REG_INTER_FRAME_PRESENCE_TIMEOUT     = 0x0041
REG_INTRA_DETECTION_ENABLED          = 0x0043
REG_INTER_DETECTION_ENABLED          = 0x0044
REG_FRAME_RATE                       = 0x0045
REG_INTRA_DETECTION_THRESHOLD        = 0x0046
REG_INTER_DETECTION_THRESHOLD        = 0x0047
REG_INTER_FRAME_DEVIATION_TIME_CONST = 0x0048
REG_INTER_FRAME_FAST_CUTOFF          = 0x0049
REG_INTER_FRAME_SLOW_CUTOFF          = 0x004A
REG_INTRA_FRAME_TIME_CONST           = 0x004B
REG_INTRA_OUTPUT_TIME_CONST          = 0x004C
REG_INTER_OUTPUT_TIME_CONST          = 0x004D
REG_AUTO_PROFILE_ENABLED             = 0x004E
REG_AUTO_STEP_LENGTH_ENABLED         = 0x004F

REG_MANUAL_PROFILE                   = 0x0050
REG_MANUAL_STEP_LENGTH               = 0x0051
REG_START                            = 0x0052
REG_END                              = 0x0053
REG_RESET_FILTERS_ON_PREPARE         = 0x0054
REG_HWAAS                            = 0x0055
REG_AUTOMATIC_SUBSWEEPS              = 0x0056
REG_SIGNAL_QUALITY                   = 0x0057

REG_DETECTION_ON_GPIO                = 0x0080

REG_COMMAND                          = 0x0100
REG_APPLICATION_ID                   = 0xFFFF


PROTOCOL_STATE_ERROR   = 0x00000001
PACKET_LENGTH_ERROR    = 0x00000002
ADDRESS_ERROR          = 0x00000004
WRITE_FAILED           = 0x00000008
WRITE_TO_READ_ONLY     = 0x00000010


RSS_REGISTER_OK           = (1 << 0)
CONFIG_CREATE_OK          = (1 << 1)
SENSOR_CREATE_OK          = (1 << 2)
DETECTOR_CREATE_OK        = (1 << 3)
DETECTOR_BUFFER_OK        = (1 << 4)
SENSOR_BUFFER_OK          = (1 << 5)
CALIBRATION_BUFFER_OK     = (1 << 6)
CONFIG_APPLY_OK           = (1 << 7)
SENSOR_CALIBRATE_OK       = (1 << 8)
DETECTOR_CALIBRATE_OK     = (1 << 9)

RSS_REGISTER_ERROR        = (1 << 16)
CONFIG_CREATE_ERROR       = (1 << 17)
SENSOR_CREATE_ERROR       = (1 << 18)
DETECTOR_CREATE_ERROR     = (1 << 19)
DETECTOR_BUFFER_ERROR     = (1 << 20)
SENSOR_BUFFER_ERROR       = (1 << 21)
CALIBRATION_BUFFER_ERROR  = (1 << 22)
CONFIG_APPLY_ERROR        = (1 << 23)
SENSOR_CALIBRATE_ERROR    = (1 << 24)
DETECTOR_CALIBRATE_ERROR  = (1 << 25)
DETECTOR_ERROR            = (1 << 28)

DETECTOR_BUSY             = (1 << 31)

ALL_ERROR = (
    RSS_REGISTER_ERROR |
    CONFIG_CREATE_ERROR |
    SENSOR_CREATE_ERROR |
    DETECTOR_CREATE_ERROR |
    DETECTOR_BUFFER_ERROR |
    SENSOR_BUFFER_ERROR |
    CALIBRATION_BUFFER_ERROR |
    CONFIG_APPLY_ERROR |
    SENSOR_CALIBRATE_ERROR |
    DETECTOR_CALIBRATE_ERROR |
    DETECTOR_ERROR |
    DETECTOR_BUSY
)


PRESENCE_DETECTED        = 0x01
PRESENCE_DETECTED_STICKY = 0x02
RESULT_DETECTOR_ERROR    = (1 << 15)
TEMPERATURE              = 0xFFFF0000


PROFILE1 = 1
PROFILE2 = 2
PROFILE3 = 3
PROFILE4 = 4
PROFILE5 = 5


APPLY_CONFIGURATION = 1
START_DETECTOR      = 2
STOP_DETECTOR       = 3

ENABLE_UART_LOGS    = 32
DISABLE_UART_LOGS   = 33
LOG_CONFIGURATION  = 34

RESET_MODULE        = 1381192737


DISTANCE_DETECTOR = 1
PRESENCE_DETECTOR = 2
REF_APP_BREATHING = 3
EXAMPLE_CARGO     = 4


class A121_Presence_Detector:
    def __init__(self, bus=0, addr=0x52, BUSY_PIN=18):
        self.i2c  = I2C(bus, scl=Pin(16), sda=Pin(15))
        self.addr = addr
        self.busy_pin = Pin(BUSY_PIN, Pin.IN)

    # ===============================
    # Write: 16-bit reg + 32-bit data
    # ===============================
    def write_u32(self, reg, data):
        data &= 0xFFFFFFFF
        buf = bytearray(6)
        buf[0] = (reg >> 8) & 0xFF
        buf[1] = reg & 0xFF
        buf[2] = (data >> 24) & 0xFF
        buf[3] = (data >> 16) & 0xFF
        buf[4] = (data >> 8) & 0xFF
        buf[5] = data & 0xFF
        self.i2c.writeto(self.addr, buf)

    # ===============================
    # Read: 16-bit reg → 32-bit data
    # ===============================
    def read_u32(self, reg):
        buf = bytearray(2)
        buf[0] = (reg >> 8) & 0xFF
        buf[1] = reg & 0xFF
        self.i2c.writeto(self.addr, buf)
        data = self.i2c.readfrom(self.addr, 4)
        return (data[0] << 24) | (data[1] << 16) | (data[2] << 8) | data[3]

    def set_start_and_end_range(self, start, end):
        self.write_u32(REG_START, start)
        self.write_u32(REG_END, end)

    def set_sweeps_per_frame(self, value):
        self.write_u32(REG_SWEEPS_PER_FRAME, value)

    def set_inter_frame_presence_timeout(self, value):
        self.write_u32(REG_INTER_FRAME_PRESENCE_TIMEOUT, value)

    def set_intra_detection_enabled(self, value):
        self.write_u32(REG_INTRA_DETECTION_ENABLED, value)

    def set_inter_detection_enabled(self, value):
        self.write_u32(REG_INTER_DETECTION_ENABLED, value)

    def set_frame_rate(self, value):
        self.write_u32(REG_FRAME_RATE, value)

    def set_intra_detection_threshold(self, value):
        self.write_u32(REG_INTRA_DETECTION_THRESHOLD, value)

    def set_inter_detection_threshold(self, value):
        self.write_u32(REG_INTER_DETECTION_THRESHOLD, value)

    def set_inter_frame_deviation_time_const(self, value):
        self.write_u32(REG_INTER_FRAME_DEVIATION_TIME_CONST, value)

    def set_inter_frame_fast_cutoff(self, value):
        self.write_u32(REG_INTER_FRAME_FAST_CUTOFF, value)

    def set_inter_frame_slow_cutoff(self, value):
        self.write_u32(REG_INTER_FRAME_SLOW_CUTOFF, value)

    def set_intra_frame_time_const(self, value):
        self.write_u32(REG_INTRA_FRAME_TIME_CONST, value)

    def set_intra_output_time_const(self, value):
        self.write_u32(REG_INTRA_OUTPUT_TIME_CONST, value)

    def set_inter_output_time_const(self, value):
        self.write_u32(REG_INTER_OUTPUT_TIME_CONST, value)

    def set_auto_profile_enabled(self, value):
        self.write_u32(REG_AUTO_PROFILE_ENABLED, value)

    def set_auto_step_length_enabled(self, value):
        self.write_u32(REG_AUTO_STEP_LENGTH_ENABLED, value)

    def set_manual_profile(self, value):
        self.write_u32(REG_MANUAL_PROFILE, value)

    def set_manual_step_length(self, value):
        self.write_u32(REG_MANUAL_STEP_LENGTH, value)

    def set_reset_filters_on_prepare(self, value):
        self.write_u32(REG_RESET_FILTERS_ON_PREPARE, value)

    def set_hwaas(self, value):
        self.write_u32(REG_HWAAS, value)

    def set_automatic_subsweeps(self, value):
        self.write_u32(REG_AUTOMATIC_SUBSWEEPS, value)

    def set_signal_quality(self, value):
        self.write_u32(REG_SIGNAL_QUALITY, value)

    def set_command(self, value):
        self.write_u32(REG_COMMAND, value)

    def get_version(self):
        v = self.read_u32(REG_VERSION)
        print(f"Major: {v >> 16:#x} Minor: {(v >> 8) & 0xff:#x} Patch: {v & 0xff:#x}")
        return v

    def get_protocol_status(self):
        return self.read_u32(REG_PROTOCOL_STATUS)

    def get_measure_counter(self):
        return self.read_u32(REG_MEASURE_COUNTER)

    def get_detector_status(self):
        return self.read_u32(REG_DETECTOR_STATUS)

    def get_presence_result(self):
        return self.read_u32(REG_PRESENCE_RESULT)

    def get_presence_distance(self):
        return self.read_u32(REG_PRESENCE_DISTANCE)

    def get_intra_presence_score(self):
        return self.read_u32(REG_INTRA_PRESENCE_SCORE)

    def get_inter_presence_score(self):
        return self.read_u32(REG_INTER_PRESENCE_SCORE)

    def get_presence_actual_frame_rate(self):
        return self.read_u32(REG_PRESENCE_ACTUAL_FRAME_RATE)

    def get_application_id(self):
        return self.read_u32(REG_APPLICATION_ID)

    def init(self):
        # wait BUSY
        while self.busy_pin.value() == 0:
            time.sleep(0.001)

        app_id = self.get_application_id()
        print(f"A121_Get_Application_Id: 0x{app_id:x}")

        self.set_command(RESET_MODULE)
        time.sleep(0.1)

        self.set_command(ENABLE_UART_LOGS)

        while self.busy_pin.value() == 0:
            time.sleep(0.001)

        self.set_start_and_end_range(300, 2500)
        self.set_frame_rate(12 * 1000)
        self.set_sweeps_per_frame(16)

        if AUTOMATIC_SUBSWEEPS:
            self.set_automatic_subsweeps(AUTOMATIC_SUBSWEEPS)
            self.set_signal_quality(20 * 1000)
        else:
            self.set_automatic_subsweeps(0)
            self.set_hwaas(32)
            if OVERRIDE:
                self.set_manual_profile(PROFILE1)
                self.set_manual_step_length(24)

        if INTRA_MOTION_DETECTION:
            self.set_intra_detection_threshold(1300)
            self.set_intra_frame_time_const(150)
            self.set_intra_output_time_const(300)

        if INTER_MOTION_DETECTION:
            self.set_inter_detection_threshold(1000)
            self.set_inter_frame_fast_cutoff(6000)
            self.set_inter_frame_slow_cutoff(200)
            self.set_inter_frame_deviation_time_const(500)
            self.set_inter_output_time_const(2000)
            self.set_inter_frame_presence_timeout(3)

        self.set_command(APPLY_CONFIGURATION)

        while True:
            status = self.get_detector_status()
            print(f"A121_Get_Detector_Status: 0x{status:x}")
            if (status & ALL_ERROR) == 0:
                break
            time.sleep(0.1)

        print("A121 init OK")

    def get_presence_detector(self):
        self.set_command(START_DETECTOR)

        while True:
            status = self.get_detector_status()
            if (status & ALL_ERROR) == 0:
                break
            time.sleep(0.01)

        result = self.get_presence_result()

        if (result & RESULT_DETECTOR_ERROR) == 0:
            if result & PRESENCE_DETECTED:
                dist = self.get_presence_distance()
                print(f"Presence_Distance: {dist} mm")

                intra = self.get_intra_presence_score()
                print(f"Intra_Presence_Score: {intra}")

                inter = self.get_inter_presence_score()
                print(f"Inter_Presence_Score: {inter}")

                fps = self.get_presence_actual_frame_rate()
                print(f"Actual_Frame_Rate: {fps}\n")

            elif result & PRESENCE_DETECTED_STICKY:
                print("PRESENCE_DETECTED_STICKY")
            else:
                print("NO PRESENCE DETECTED")
        else:
            print("A121 DETECTOR ERROR (need RESET_MODULE)")


if __name__ == "__main__":
    a121 = A121_Presence_Detector()
    a121.init()

    try:
        while True:
            a121.get_presence_detector()
            time.sleep(0.02)

    except KeyboardInterrupt:
        print("Quit.")

